File: backend/src/models/Playlist.js
const mongoose = require('mongoose');

const playlistSchema = new mongoose.Schema({
  userId: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User',
    required: true,
    index: true,
  },
  title: {
    type: String,
    required: true,
    trim: true,
    maxlength: 100,
    index: true,
  },
  description: {
    type: String,
    trim: true,
    maxlength: 500,
    default: '',
  },
  coverImageUrl: {
    type: String,
    default: null,
  },
  songIds: [{
    type: String,
    trim: true,
  }],
  totalDuration: {
    type: Number,
    default: 0,
  },
  isPublic: {
    type: Boolean,
    default: true,
  },
  followerCount: {
    type: Number,
    default: 0,
  },
  playCount: {
    type: Number,
    default: 0,
  },
}, {
  timestamps: true,
});

playlistSchema.index({ userId: 1, createdAt: -1 });

playlistSchema.methods.toPublicJSON = function() {
  return {
    id: this._id,
    userId: this.userId,
    title: this.title,
    description: this.description,
    coverImageUrl: this.coverImageUrl,
    songIds: this.songIds,
    totalDuration: this.totalDuration,
    isPublic: this.isPublic,
    followerCount: this.followerCount,
    playCount: this.playCount,
    createdAt: this.createdAt,
    updatedAt: this.updatedAt,
  };
};

module.exports = mongoose.model('Playlist', playlistSchema);

File: lib/models/playlist.dart
import 'song.dart';

class Playlist {
  final String id;
  final String? userId;
  final String title;
  final String description;
  final String? coverImageUrl;
  final List<String> songIds;
  final int totalDuration;
  final bool isPublic;
  final int followerCount;
  final int playCount;
  final DateTime? createdAt;
  final List<Song> songs;

  const Playlist({
    required this.id,
    this.userId,
    required this.title,
    required this.description,
    this.coverImageUrl,
    required this.songIds,
    required this.totalDuration,
    required this.isPublic,
    required this.followerCount,
    required this.playCount,
    this.createdAt,
    this.songs = const [],
  });

  factory Playlist.fromJson(Map<String, dynamic> json) {
    return Playlist(
      id: json['id'] as String? ?? (json['_id'] as String? ?? ''),
      userId: json['userId'] as String?,
      title: json['title'] as String? ?? 'Untitled',
      description: json['description'] as String? ?? '',
      coverImageUrl: json['coverImageUrl'] as String?,
      songIds: (json['songIds'] as List<dynamic>? ?? []).cast<String>(),
      totalDuration: (json['totalDuration'] as num?)?.toInt() ?? 0,
      isPublic: json['isPublic'] as bool? ?? true,
      followerCount: (json['followerCount'] as num?)?.toInt() ?? 0,
      playCount: (json['playCount'] as num?)?.toInt() ?? 0,
      createdAt: json['createdAt'] != null
          ? DateTime.tryParse(json['createdAt'] as String)
          : null,
      songs: (json['songs'] as List<dynamic>? ?? [])
          .whereType<Map<String, dynamic>>()
          .map(Song.fromJson)
          .toList(),
    );
  }
}